File: src/main/java/com/example/finsentinel/controller/ResearchController.java
package com.example.finsentinel.controller;

import com.example.finsentinel.dto.research.CompanyProfile;
import com.example.finsentinel.dto.research.FinancialMetrics;
import com.example.finsentinel.service.research.CompanyResearchService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST endpoints for company research data (profile, financials).
 *
 * <p>Delegates to {@link CompanyResearchService} which fetches data via the
 * configured {@link com.example.finsentinel.service.research.ResearchDataProvider}.
 */
@RestController
@RequestMapping("/api/research")
@RequiredArgsConstructor
public class ResearchController {

    private final CompanyResearchService companyResearchService;

    @GetMapping("/profile/{ticker}")
    public ResponseEntity<CompanyProfile> getProfile(@PathVariable String ticker) {
        CompanyProfile profile = companyResearchService.getCompanyProfile(ticker);
        if (profile == null) return ResponseEntity.notFound().build();
        return ResponseEntity.ok(profile);
    }

    @GetMapping("/financials/{ticker}")
    public ResponseEntity<List<FinancialMetrics>> getFinancials(
            @PathVariable String ticker,
            @RequestParam(defaultValue = "4") int periods) {
        return ResponseEntity.ok(companyResearchService.getFinancialMetrics(ticker, periods));
    }
}

File: apps/api/src/research/research.controller.ts
import {
  Controller,
  Get,
  Param,
  Query,
  UseGuards,
} from '@nestjs/common';
import { JwtGuard } from '../auth/jwt.guard';
import { CompanyResearchService } from './company-research.service';
import { parseIntParam } from '../common/utils/parse-int-param';

/**
 * Research controller — company profile and financial metrics.
 */
@Controller('research')
@UseGuards(JwtGuard)
export class ResearchController {
  constructor(private readonly researchService: CompanyResearchService) {}

  /** GET /research/profile/:ticker — company profile. */
  @Get('profile/:ticker')
  async getProfile(
    @Param('ticker') ticker: string,
    @Query('provider') provider?: string,
  ) {
    return this.researchService.getCompanyProfile(ticker, provider);
  }

  /** GET /research/financials/:ticker — financial metrics. */
  @Get('financials/:ticker')
  async getFinancials(
    @Param('ticker') ticker: string,
    @Query('periods') periodsParam?: string,
    @Query('provider') provider?: string,
  ) {
    const periods = parseIntParam(periodsParam, 4, 1, 20);
    return this.researchService.getFinancialMetrics(ticker, periods, provider);
  }
}
